marathon_acme/sse_protocol.py

```diff
@@ -32,10 +32,29 @@ def when_finished(self):
         """
         Get a deferred that will be fired when the connection is closed.
         """
-        d = Deferred()
+        d = Deferred(canceller=self._cancel)
         self._waiting.append(d)
         return d
 
+    def _cancel(self, d):
+        self.log.warn('SSE connection cancelled')
+        # Remove the deferred from the waiting list-- can't call it back, it's
+        # been cancelled.
+        self._waiting.remove(d)
+        self._lose_connection()
+
+    def _lose_connection(self):
+        assert self.transport is not None
+        self.log.warn('Purposefully losing SSE connection...')
+
+        # HACK: We don't really know what our transport will be. We could be
+        # given an ITransport (with loseConnection()) or an IPushProducer (with
+        # stopProducing()).
+        if hasattr(self.transport, 'loseConnection'):
+            self.transport.loseConnection()
+        elif hasattr(self.transport, 'stopProducing'):
+            self.transport.stopProducing()
+
```

Comment:
Could throw a `ValueError` in [`makeConnection`](http://twistedmatrix.com/documents/current/api/twisted.internet.protocol.BaseProtocol.html#makeConnection)